### PR TITLE
chore: 🍰 Refactor Contribution Form Logic And Write Tests

### LINKED ISSUE #2091: 🔧 [Refactor] Refactor Contribution Form Logic And Write Tests
## 🔧 Refactor ticket

Refactor contribution form logic and write tests.

- refine logic of `lastMonthObject` etc.
- check year in `isThisMonth`
- test it

### PR DESCRIPTION
**Pull request by Tirokk**
## 🍰 Pullrequest

Refactor contribution form logic and write tests.

### Issues

- fixes #2091

### Todo

- [x] refine logic of `lastMonthObject` etc.
- [x] check year in `isThisMonth`
- [x] test it
  - [x] partly
- [x] test for memo length (use constant)
  - [x] < 5
  - [x] >= 5 <= 255
  - [x] > 255
- [x] cleanup: search for `Wolle`

### CODE REVIEW

File: admin/jest.config.js [hunk 1]
```diff
@@ -26,5 +26,5 @@ module.exports = {
   testMatch: ['**/?(*.)+(spec|test).js?(x)'],
   // snapshotSerializers: ['jest-serializer-vue'],
   transformIgnorePatterns: ['<rootDir>/node_modules/(?!vee-validate/dist/rules)'],
-  testEnvironment: 'jest-environment-jsdom-sixteen',
+  // testEnvironment: 'jest-environment-jsdom-sixteen', // not needed anymore since jest@26, see: https://www.npmjs.com/package/jest-environment-jsdom-sixteen
```
Comment (Tirokk) [inline admin/jest.config.js:29]: I allowed my self to remove this.

File: admin/package.json [hunk 2]
```diff
@@ -70,7 +70,6 @@
     "eslint-plugin-prettier": "3.3.1",
     "eslint-plugin-promise": "^5.1.1",
     "eslint-plugin-vue": "^7.20.0",
-    "jest-environment-jsdom-sixteen": "^2.0.0",
```
Comment (Tirokk) [inline admin/package.json:73]: same same

File: frontend/jest.config.js [hunk 3]
```diff
@@ -23,5 +23,5 @@ module.exports = {
   testMatch: ['**/?(*.)+(spec|test).js?(x)'],
   // snapshotSerializers: ['jest-serializer-vue'],
   transformIgnorePatterns: ['<rootDir>/node_modules/(?!vee-validate/dist/rules)'],
-  testEnvironment: 'jest-environment-jsdom-sixteen',
+  // testEnvironment: 'jest-environment-jsdom-sixteen', // not needed anymore since jest@26, see: https://www.npmjs.com/package/jest-environment-jsdom-sixteen
```
Comment (Tirokk) [inline frontend/jest.config.js:26]: I allowed my self to remove this.
Especally in the hope it makes my date tests working with `jest
  .useFakeTimers()
  .setSystemTime()`. But it unfortunatelly did not help.

File: frontend/package.json [hunk 4]
```diff
@@ -44,7 +44,6 @@
     "identity-obj-proxy": "^3.0.0",
     "jest": "^26.6.3",
     "jest-canvas-mock": "^2.3.1",
-    "jest-environment-jsdom-sixteen": "^2.0.0",
```
Comment (Tirokk) [inline frontend/package.json:47]: same same

File: frontend/src/components/Contributions/ContributionForm.spec.js [hunk 5]
```diff
@@ -48,64 +48,365 @@ describe('ContributionForm', () => {
...
+      })
+
+      // jest-environment-jsdom-sixteen
+      // testEnvironment: 'jest-environment-jsdom-sixteen',
+      // https://www.npmjs.com/package/jest-environment-jsdom-sixteen
+      // https://dev.to/philw_/problems-with-using-usefaketimersmodern-in-a-create-react-app-cra-project-with-jest-26-and-lodashs-debounce-function-3ohd
+      describe.skip('date in middle of year', () => {
+        describe('same month', () => {
+          beforeEach(async () => {
+            jest.useFakeTimers('modern')
```
Comment (Tirokk) [inline frontend/src/components/Contributions/ContributionForm.spec.js:112]: Here I get the error message:

`TypeError: Cannot read properties of undefined (reading 'useFakeTimers')`

Here is an hint why. But it didn't help after I followed the suggestion. Unfortunatelly …

File: frontend/src/components/Contributions/ContributionForm.spec.js [hunk 6]
```diff
@@ -48,64 +48,365 @@ describe('ContributionForm', () => {
...
+              expect(wrapper.vm.isThisMonth).toBe(false)
+            })
+          })
+        })
+      })
+
+      describe.skip('date in january', () => {
+        describe('same month', () => {
+          beforeEach(async () => {
+            jest.useFakeTimers('modern').setSystemTime(new Date('2020-01-12'))
```
Comment (Tirokk) [inline frontend/src/components/Contributions/ContributionForm.spec.js:145]: same same

File: frontend/src/components/Contributions/ContributionForm.vue [hunk 7]
```diff
@@ -104,11 +104,11 @@ export default {
...
-      if (this.form.id) {
-        this.$emit('update-contribution', this.form)
-      } else {
-        this.$emit('set-contribution', this.form)
-      }
+      // not working for testing:
+      // this.$emit(this.form.id ? 'update-contribution' : 'set-contribution', this.form)
+      // works for testing:
+      // why ever, we have to make a spread '...', to evaluate the values it looks like: (I didn't find a solution in the test itmself)
+      this.$emit(this.form.id ? 'update-contribution' : 'set-contribution', { ...this.form })
```
Comment (Tirokk) [inline frontend/src/components/Contributions/ContributionForm.vue:111]: Now we have a solution to submit in tests and check for the emitted values. 🍀
YEAH !!!

File: frontend/src/components/Contributions/ContributionForm.vue [hunk 8]
```diff
@@ -122,53 +122,48 @@ export default {
...
-    /*
-     * minimalDate() = Sets the date to the 1st of the previous month.
-     *
-     */
+    // sets the date to the 1st of the previous month
     minimalDate() {
-      return new Date(this.maximalDate.getFullYear(), this.maximalDate.getMonth() - 1, 1)
+      const month = this.maximalDate.getMonth()
+      const year = this.maximalDate.getFullYear()
+      return new Date(year + (month === 0 ? -1 : 0), month === 0 ? 11 : month - 1, 1)
```
[the remark this answers is not in the document]
Comment (Mogge) [reply frontend/src/components/Contributions/ContributionForm.vue:138]: I recommend to try such things in the browser's console

File: frontend/src/components/Contributions/ContributionForm.spec.js (hunk 5 above)
Comment (Tirokk) [reply frontend/src/components/Contributions/ContributionForm.spec.js:108]: Did followup issue #2128

File: frontend/jest.config.js (hunk 3 above)
Comment (Tirokk) [reply frontend/jest.config.js:26]: Did followup issue #2128

File: frontend/src/components/Contributions/ContributionForm.vue (hunk 8 above)
[the remark this answers is not in the document]
Comment (Tirokk) [reply frontend/src/components/Contributions/ContributionForm.vue:138]: We need the first of the month …

File: frontend/src/components/Contributions/ContributionForm.vue (hunk 7 above)
Comment (Tirokk) [reply frontend/src/components/Contributions/ContributionForm.vue:115]: So what else can we do right or better?

Comment (Tirokk) [reply frontend/src/components/Contributions/ContributionForm.vue:115]: I did it now with @Mogge and @Elweyn

File: frontend/src/components/Contributions/ContributionForm.vue (hunk 8 above)
[the remark this answers is not in the document]
Comment (Tirokk) [reply frontend/src/components/Contributions/ContributionForm.vue]: I did it now with @ogerly  and @Elweyn

File: frontend/src/components/Contributions/ContributionForm.vue (hunk 7 above)
Comment (Mogge) [reply frontend/src/components/Contributions/ContributionForm.vue:115]: Looks fine. Please remove the comments

File: frontend/src/components/Contributions/ContributionForm.vue [hunk 9]
```diff
@@ -122,53 +122,47 @@ export default {
...
-    /*
-     * minimalDate() = Sets the date to the 1st of the previous month.
-     *
-     */
     minimalDate() {
-      return new Date(this.maximalDate.getFullYear(), this.maximalDate.getMonth() - 1, 1)
+      // sets the date to the 1st of the previous month
+      let date = new Date(this.maximalDate) // has to be a new object, because of 'setMonth' changes the objects date
+      date = new Date(date.setMonth(date.getMonth() - 1))
+      return new Date(date.getFullYear(), date.getMonth(), 1)
```
Comment (Mogge) [inline frontend/src/components/Contributions/ContributionForm.vue:138]: The day does not matter at all. But it is okay as it is

**Review by Mogge — DISMISSED** [review]

**Review by ogerly — CHANGES_REQUESTED** [review]

File: frontend/src/components/Contributions/ContributionForm.spec.js [hunk 10]
```diff
@@ -48,64 +48,399 @@ describe('ContributionForm', () => {
     describe('empty form data', () => {
       describe('buttons', () => {
         it('has reset enabled', () => {
```
[the remark this answers is not in the document]
Comment (Tirokk) [reply frontend/src/components/Contributions/ContributionForm.spec.js]: Done

**Review by Mogge — APPROVED** [review]

File: frontend/src/components/Contributions/ContributionForm.vue (hunk 7 above)
Comment (Tirokk) [reply frontend/src/components/Contributions/ContributionForm.vue:112]: I reduced the comment to a minimum …

**Review by Mogge — APPROVED** [review]

**Pull request merged**